### DISCUSSION TITLE
Passing through an existing Qtile keybinding into a specific application

### DISCUSSION THREAD

**Discussion by raj-magesh** [qtile/qtile#3037] [Q&A]
I use a terminal (kitty) with tiling-like support and I'd like to reuse the same keybindings I use in Qtile with the kitty panes. However, Qtile has priority over the keybindings. I've managed to disable certain Qtile keybindings whenever a terminal window is focused using this minimal example:

```
def move_focus(qtile, direction):
    if qtile.current_window.get_wm_class()[0] != 'kitty':
        if direction == 'left':
            qtile.current_group.layout.cmd_left()
    else:
        qtile.cmd_simulate_keypress([MOD], 'h')

keys = [Key([MOD], 'h', lazy.function(move_focus, direction='left'), desc='move focus left')]
```

However, the key combination still doesn't get passed through to the terminal. How can I achieve that?

I've found other people with a similar issue:
- https://groups.google.com/g/qtile-dev/c/cp83eFM-JwQ
- https://groups.google.com/g/qtile-dev/c/e25bgej9Ses/m/pAczjD9lmEoJ

**Comment by m-col** [comment]
I don't think this is possible currently.

There is a work-in-progress PR that aims to make key swallowing configurable: https://github.com/qtile/qtile/pull/2925

I'm not sure that will wholly get what you want, however. You would also need to let `LazyCall.when` accept a `Match` object to determine whether or not the keybinding should be executed. Nobody is working on this but if you're interested it would make a great feature, and would probably be pretty easy to implement.

**Reply by m-col** [reply]
Then you would just need to do:
```py
Key(mods, key, lazy.layout.left().when(match=Match(wm_class="kitty")))
```
Though actually, looking at it, it would probably need a `not_match` to blacklist windows with the match, rather than whitelist, which is what this looks like.

**Reply by elParaguayo** [reply]
You might want both. Whitelist for the action to only apply to matching windows and a blacklist for the action to apply to all windows other than those matching the blacklist.

**Reply by raj-magesh** [reply]
Thanks for the response! @elParaguayo mentioned the same PR in another discussion thread and I'll be happy to test it later.

I also don't see why my solution won't work.
It doesn't use a lazy call but I think it should work if the keypress isn't swallowed? Is there a reason the snippet above would fail?

I'd be happy to try and work on making the LazyCall.when() accept a Match input (but I'm new to Qtile, so I have a steep learning curve ahead.)

**Reply by m-col** [reply]
The `lazy.function` used in your keybinding creates a lazy call that will later execute the function you've passed it. In that case you could do `lazy.function(move_focus, "left").when(<criteria>)`. This swallows the keybinding. Then the `qtile.cmd_simulate_keypress([MOD], 'h')` is being used in a way that it wasn't really intended to be used - and that's on us for not documenting it correctly. That command is only used for internal bindings, and keys passed to it are never passed to a client. This has come up before, and there isn't really any reason why it couldn't do that. I think the command was originally used for testing.

**Reply by raj-magesh** [reply]
That makes sense, thanks for the explanation!

**Reply by jwijenbergh** [reply]
> Then you would just need to do:
> 
> ```python
> Key(mods, key, lazy.layout.left().when(match=Match(wm_class="kitty")))
> ```
> 
> Though actually, looking at it, it would probably need a `not_match` to blacklist windows with the match, rather than whitelist, which is what this looks like.

Indeed! I commented this exact thing on the issue as I didn't see the discussion here. Do we also maybe want to support arbitrary lazy functions in ```.when()```? That way users can make their own completely customizable conditions.
